public/app/features/explore/state/actions.ts

```diff
@@ -530,6 +533,124 @@ export const runQueries = (exploreId: ExploreId): ThunkResult<void> => {
   };
 };
 
+/**
+ * Sub-action to run append queries
+ */
+export const runAppendQueries = (exploreId: ExploreId, direction: QueryDirection): ThunkResult<void> => {
+  return (dispatch, getState) => {
+    dispatch(updateTime({ exploreId }));
+
+    const exploreItemState = getState().explore[exploreId];
+    const {
+      datasourceInstance,
+      queries,
+      containerWidth,
+      isLive: live,
+      range,
+      scanning,
+      queryResponse,
+      querySubscription,
+      mode,
+      showingGraph,
+      showingTable,
+    } = exploreItemState;
+
+    const newestResponseEndTimestamp = queryResponse.series[0].fields[0].values.get(
+      Math.floor((queryResponse.series[0].fields[0].values.length - 1) / 2)
+    );
+
+    const oldestResponseEndTimestamp = queryResponse.series[queryResponse.series.length - 1].fields[0].values.get(
+      Math.floor((queryResponse.series[0].fields[0].values.length - 1) / 2)
+    );
+
+    const newRaw: RawTimeRange = {
+      from:
+        direction === QueryDirection.forward
+          ? processDateTime(newestResponseEndTimestamp)
+          : processDateTime(range.raw.from),
+      to:
+        direction === QueryDirection.backward
+          ? processDateTime(oldestResponseEndTimestamp)
+          : processDateTime(range.raw.to),
+    };
```

Comment:
Would be probably more readable as:
```
if (direction === QueryDirection.forward) {
  // build time range
} else {
  // build time range
}
```

Reply:
Right, will push the refactored version asap.